packages/framework/attributor/src/mixinAttributor.ts

```diff
@@ -0,0 +1,216 @@
+            await runtime.runtimeAttributor.initialize(
+                deltaManager,
+                audience,
+                baseSnapshot,
+                async (id) => runtime.storage.readBlob(id)
+            );
+            return runtime;
+        }
+
+        private runtimeAttributor: RuntimeAttributor | undefined;
+
+        protected addContainerStateToSummary(
+            summaryTree: ISummaryTreeWithStats,
+            fullTree: boolean,
+            trackState: boolean,
+            telemetryContext?: ITelemetryContext,
+        ) {
+            super.addContainerStateToSummary(summaryTree, fullTree, trackState, telemetryContext);
+            const attributorSummary = this.runtimeAttributor?.summarize();
+            if (attributorSummary) {
+                addSummarizeResultToSummary(summaryTree, attributorKey, attributorSummary);
+            }
+        }
+    } as unknown as typeof ContainerRuntime;
+
+
+class RuntimeAttributor implements IRuntimeAttributor {
+    public get IRuntimeAttributor(): IRuntimeAttributor { return this; };
+
+    public get(key: AttributionKey): AttributionInfo {
+        assert(this.opAttributor !== undefined,
+            "RuntimeAttributor must be initialized before getAttributionInfo can be called");
+        
+        return this.opAttributor.getAttributionInfo(key.seq);
+    }
+
+    public has(key: AttributionKey): boolean {
+        return this.opAttributor?.tryGetAttributionInfo(key.seq) !== undefined;
+    }
+    
+    private encoder: Encoder<IAttributor, string> = {
+        encode: unreachableCase,
+        decode: unreachableCase
+    };
+
+
+    private opAttributor: IAttributor | undefined;
+    private skipSummary = true;
+
+    public async initialize(
+        deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>,
+        audience: IAudience,
+        baseSnapshot: ISnapshotTree | undefined,
+        readBlob: (id: string) => Promise<ArrayBufferLike>,
+    ): Promise<void> {
+        const snapshot = baseSnapshot?.trees[attributorKey];
+        // Existing documents that don't already have a snapshot containing runtime attribution info shouldn't
+        // inject any for now--this causes some back-compat integration problems that aren't fully worked out.
+        const shouldExcludeAttributor = baseSnapshot !== undefined && snapshot === undefined;
+        if (shouldExcludeAttributor) {
+            // This gives a consistent error for calls to `get` on keys that don't exist.
+            this.opAttributor = new Attributor();
+            return;
+        }
+
+        this.skipSummary = false;
+        this.encoder = chain(
+            new AttributorSerializer(
+                (entries) => new OpStreamAttributor(deltaManager, audience, entries),
+                deltaEncoder
+            ),
+            makeLZ4Encoder(),
+        );
+
+        if (snapshot !== undefined) {
+            const opAttributorTree = snapshot.trees[opKey];
```

Comment:
i'm a bit worried about this. is there an example of where this goes? will it conflict with anything? should the `op` tree be nested under and `attribution' tree

Reply:
ahh. i see above it is nested. maybe rename to attributorSnapshot or something. still not totally clear to me where it lands in the snapshot.

Reply:
you can infer it from the tests a bit, but snapshot format on actual documents will look something like
```
{
    ".channels": { ... },
    ".protocol": { ... },
    "attributor": {
        "trees": {
            "op": { "blobs": { "attributor": /* id of blob uploaded with contents of encoder.encode(opAttributor) */ } } 
        }
    }
}
```

now that I think about it, it's probably fine to leave the space at the blob level and condense this a bit to

```
{
    ".channels": { ... },
    ".protocol": { ... },
    "attributor": {
        "blobs": {
            "op": /* id of blob uploaded with contents of encoder.encode(opAttributor) */ 
        }
    }
}
```

Reply:
i think i've convinced myself this will be ok. the `attributor` tree should be a peer of the `.channels` tree

Reply:
just seeing replies not. wish they auto updated. but sounds good.